File: src/542-01-matrix/solution.rb
# @param {Integer[][]} mat
# @return {Integer[][]}
def update_matrix(mat)
  rows, cols = mat.length, mat.first.length
  res = Array.new(rows) { Array.new(cols) { Float::INFINITY } } 
  0.upto(rows-1) do |i|
    0.upto(cols-1) do |j|
      if mat[i][j] == 0
        res[i][j] = 0
      else
        if i > 0
          res[i][j] = [res[i][j], res[i - 1][j] + 1].min
        end
        
        if j > 0
          res[i][j] = [res[i][j], res[i][j - 1] + 1].min
        end
      end
    end
  end

  (rows-1).downto(0) do |i|
    (cols-1).downto(0) do |j|
      if mat[i][j] == 0
        res[i][j] = 0
      else
        if i < rows - 1
          res[i][j] = [res[i][j], res[i + 1][j] + 1].min
        end
        
        if j < cols - 1
          res[i][j] = [res[i][j], res[i][j + 1] + 1].min
        end
      end
    end
  end
  
  res
end

File: src/542-01-matrix/solution.ts
function updateMatrix(mat: number[][]): number[][] {
  let m: number = mat.length, n: number = mat[0].length;
  let res: number[][] = Array.from({ length: m }, () => Array(n).fill(1_000_000));

  for(let i = 0; i < m; i++) {
    for(let j = 0; j < n; j++) {
      if(mat[i][j] === 0){
        res[i][j] = 0;
      } else {
        if(i > 0) res[i][j] = Math.min(res[i][j], res[i - 1][j] + 1);
        if(j > 0) res[i][j] = Math.min(res[i][j], res[i][j - 1] + 1);
      }
    }
  }

  for(let i = m - 1; i >= 0; i--) {
    for(let j = n - 1; j >= 0; j--) {
      if(mat[i][j] === 0){
        res[i][j] = 0;
      } else {
        if(i < m - 1) res[i][j] = Math.min(res[i][j], res[i + 1][j] + 1);
        if(j < n - 1) res[i][j] = Math.min(res[i][j], res[i][j + 1] + 1);
      }
    }
  }

  return res;
};
